rust/crypto/src/aes_gcm.rs

```diff
@@ -0,0 +1,272 @@
+//
+// Copyright 2021 Signal Messenger, LLC.
+// SPDX-License-Identifier: AGPL-3.0-only
+//
+
+use crate::aes::Aes256;
+use crate::ghash::Ghash;
+use crate::{Aes256Ctr32, Error, Result};
+use subtle::ConstantTimeEq;
+
+pub const TAG_SIZE: usize = 16;
+pub const NONCE_SIZE: usize = 12;
+
+struct GcmGhash {
+    ghash: Ghash,
+    nonce: [u8; TAG_SIZE],
+    msg_buf: [u8; TAG_SIZE],
+    msg_buf_offset: usize,
+    ad_len: usize,
+    msg_len: usize,
+}
+
+impl GcmGhash {
+    fn new(h: &[u8; TAG_SIZE], nonce: [u8; TAG_SIZE], associated_data: &[u8]) -> Result<Self> {
+        let mut ghash = Ghash::new(h)?;
+
+        ghash.update_padded(associated_data)?;
+
+        Ok(Self {
+            ghash,
+            nonce,
+            msg_buf: [0u8; TAG_SIZE],
+            msg_buf_offset: 0,
+            ad_len: associated_data.len(),
+            msg_len: 0,
+        })
+    }
+
+    fn update(&mut self, msg: &[u8]) -> Result<()> {
+        if self.msg_buf_offset > 0 {
+            let taking = std::cmp::min(msg.len(), TAG_SIZE - self.msg_buf_offset);
+            self.msg_buf[self.msg_buf_offset..self.msg_buf_offset + taking]
+                .copy_from_slice(&msg[..taking]);
+            self.msg_buf_offset += taking;
+            assert!(self.msg_buf_offset <= TAG_SIZE);
+
+            self.msg_len += taking;
+
+            if self.msg_buf_offset == TAG_SIZE {
+                self.ghash.update(&self.msg_buf)?;
+                self.msg_buf_offset = 0;
+                return self.update(&msg[taking..]);
+            } else {
+                return Ok(());
+            }
+        }
+
+        self.msg_len += msg.len();
+
+        assert_eq!(self.msg_buf_offset, 0);
+        let full_blocks = msg.len() / 16;
+        let leftover = msg.len() - 16 * full_blocks;
+        assert!(leftover < TAG_SIZE);
+        if full_blocks > 0 {
+            self.ghash.update(&msg[..full_blocks * 16])?;
+        }
+
+        self.msg_buf[0..leftover].copy_from_slice(&msg[full_blocks * 16..]);
+        self.msg_buf_offset = leftover;
+        assert!(self.msg_buf_offset < TAG_SIZE);
+
+        Ok(())
+    }
+
+    fn finalize(mut self) -> Result<[u8; TAG_SIZE]> {
+        if self.msg_buf_offset > 0 {
+            for i in self.msg_buf_offset..TAG_SIZE {
+                self.msg_buf[i] = 0;
+            }
+            self.ghash.update(&self.msg_buf)?;
+        }
+
+        let mut final_block = [0u8; 16];
+        final_block[..8].copy_from_slice(&(8 * self.ad_len).to_be_bytes());
+        final_block[8..].copy_from_slice(&(8 * self.msg_len).to_be_bytes());
+
+        self.ghash.update(&final_block)?;
+        let mut hash = self.ghash.finalize()?;
+
+        for (i, b) in hash.iter_mut().enumerate() {
+            *b ^= self.nonce[i];
```

Comment:
It took me a while to realize that this nonce wasn't the same as the original nonce; would `nonce_after_ctr` work instead or would that just cause more confusion / feel non-idiomatic?

Reply:
Something like `ghash_pad` is probably better as  `pad` is a somewhat common notation for this value in the general context of Carter-Wegman MACs (of which GCM's authentication function is an example).